File: src/Read2Me.App/Live/LiveMessages.cs
using Read2Me.Services.Audio;
using Read2Me.Services.BookEdits;
using Read2Me.Services.Characters;
using Read2Me.Services.Llm;
using Read2Me.Services.Mutations;
using Read2Me.Services.NodeStatus;

namespace Read2Me.App.Live;

/// <summary>
/// Wire shapes for <c>/hubs/live</c> (Angular ticket 06; mirrored in TypeScript by ticket 07). One
/// record per server→client family; a family with several event kinds carries a <c>kind</c>
/// discriminator and nullable fields so a client can switch on one string. Enums travel as their
/// member names, nulls are omitted (see <see cref="LiveServiceCollectionExtensions"/>).
/// </summary>
public static class LiveGroups
{
    public const string Global = "global";
    public const string StreamLlm = "stream:llm";
    public const string StreamAudio = "stream:audio";
    public static string Project(string folder) => $"project:{folder}";
}

/// <summary>Both queues' roll-ups plus the attribution escalation banner (debounced, group <c>global</c>).</summary>
public sealed record QueueMessage(
    QueueSnapshot Attribution,
    AudioQueueSnapshot Audio,
    EscalationState? Escalation);

/// <summary>The latched "escalating N items → config (step S)" label; null when no escalation is active.</summary>
public sealed record EscalationState(int Step, string? ConfigName, int ItemCount);

/// <summary>
/// Per-node roll-up deltas for one project (debounced, group <c>project:{folder}</c>). A null value
/// means the node no longer rolls up (the folder was cleared or re-seeded without it).
/// </summary>
public sealed record NodeStatusMessage(
    string Folder,
    IReadOnlyDictionary<string, NodeStatusSummary?> Nodes,
    int FolderAudioRemaining);

/// <summary>Per-paragraph attribution and per-item audio status deltas for one project (debounced).</summary>
public sealed record ItemStatusMessage(
    string Folder,
    IReadOnlyDictionary<string, ParagraphStatusEntry?> Paragraphs,
    IReadOnlyDictionary<string, ItemStatusEntry?> Items);

/// <summary>What the attribution queue currently says about one paragraph; null fields mean "nothing".</summary>
public sealed record ParagraphStatusEntry(ParagraphQueueStatus? Status, ParagraphOutcome? Outcome);

/// <summary>What the audio queue and the review mirror say about one item.</summary>
public sealed record ItemStatusEntry(
    AudioItemQueueStatus? Status,
    AudioItemOutcome? Outcome,
    long? AudioVersion,
    AudioReviewInfo? Review);

/// <summary><see cref="BookMutationReceipt"/> as JSON with the folder flattened; <c>originId</c> untouched.</summary>
public sealed record ReceiptMessage(
    string Folder,
    string MutationName,
    Guid MutationId,
    long Revision,
    BookMutationEffects Effects,
    Guid OriginId)
{
    public static ReceiptMessage From(BookMutationReceipt r) =>
        new(r.FolderId.Value, r.MutationName, r.MutationId, r.Revision, r.Effects, r.OriginId);
}

/// <summary>
/// <c>kind</c>: phaseStarted | progress | completed | failed | cancelled. <c>Folder</c> is the project the
/// (single, global) run belongs to; <c>OutputFileName</c> rides on <c>completed</c>.
/// </summary>
public sealed record AssemblyMessage(
    string Kind, string? Phase = null, double? Fraction = null, string? Reason = null,
    string? Folder = null, string? OutputFileName = null);

/// <summary>Assembly service state for the connect-time snapshot.</summary>
public sealed record AssemblyState(
    bool IsRunning, string? CurrentPhase, double EncodePercent, string? LastError, int AudioRemainingCount,
    string? Folder = null, string? OutputFileName = null);

/// <summary><c>kind</c>: started | progress | voiceUpdated | completed | cancelled.</summary>
public sealed record VoiceBatchMessage(
    string Kind,
    string? Operation = null,
    int? Processed = null,
    int? Total = null,
    int? Failed = null,
    string? CurrentVoiceName = null,
    Guid? CharacterId = null,
    Guid? VoiceId = null,
    string? DesignPrompt = null,
    string? AudioFileName = null,
    string? Transcript = null);

/// <summary>Voice batch runner state for the connect-time snapshot.</summary>
public sealed record VoiceBatchState(
    bool IsRunning, int Processed, int Total, int Failed, string? CurrentVoiceName, string? CurrentOperation, string? LastError);

/// <summary><c>kind</c>: recoveryStarted | containerRestarted | serviceHealthy | serviceDown.</summary>
public sealed record WatchdogMessage(string Kind, string Service, string? Reason = null);

/// <summary>
/// A managed service's status as last observed (<c>AiServiceStatus</c> member name), pushed after
/// every probe, lifecycle op and watchdog transition so chips follow without polling (Angular
/// ticket 25). <c>op</c> (start | restart | shutdown) with <c>ok</c> / <c>error</c> is present only
/// when a lifecycle op produced the observation — that is what a client toasts.
/// </summary>
public sealed record ServiceStatusMessage(
    string Name, string Status, string? Op = null, bool? Ok = null, string? Error = null);

/// <summary>
/// <c>kind</c>: stage | done — one pre-flight run (Angular ticket 25), sent to the connection that
/// started it. <c>stage</c> carries a service and its stage (waitingToStop | stopping | stopped |
/// waitingToStart | starting | ready | failed); <c>done</c> carries <c>ok</c> and, on failure, the
/// summary <c>reason</c>. <c>run</c> is the id the 202 answered with.
/// </summary>
public sealed record PreflightMessage(
    string Kind,
    string Run,
    string? Name = null,
    string? Stage = null,
    string? Error = null,
    bool? Ok = null,
    string? Reason = null);

/// <summary>
/// <c>kind</c>: runStarted | runEnded | requestStarted | delta | streamCompleted | streamFailed |
/// streamAborted | escalationStarted. <c>delta</c> is a 100 ms batch of thinking + content text.
/// Stream group only.
/// </summary>
public sealed record LlmMessage(
    string Kind,
    string? Thinking = null,
    string? Content = null,
    string? ParagraphPreview = null,
    string? Prompt = null,
    int? ConfigId = null,
    string? ConfigName = null,
    int? TokensIn = null,
    int? TokensOut = null,
    double? GenerationMs = null,
    double? TokensPerSecond = null,
    string? Reason = null,
    int? Step = null,
    int? ItemCount = null);

/// <summary>
/// <c>kind</c>: itemStarted | audioGenerated | normalized | postProcessed | transcribed | verified |
/// failed — one per <see cref="AudioGenEvent"/>. Stream group only.
/// </summary>
public sealed record AudioGenMessage(
    string Kind,
    Guid Id,
    int Attempt,
    string? Character = null,
    string? Text = null,
    bool? Ok = null,
    string? Reason = null,
    string? StepId = null,
    bool? Applied = null,
    string? Transcript = null,
    double? Wer = null,
    bool? Rescued = null);

/// <summary>Which settings area changed, so other clients refresh their list.</summary>
public sealed record SettingsChangedMessage(string Area);

/// <summary>
/// <c>kind</c>: progress | done | failed — one AI book-edit proposal run (Angular ticket 19), sent
/// to the connection that started it. <c>program</c> is the session id the run belongs to;
/// <c>done</c> carries every row the run landed (all of them, or the partial set a cancel kept).
/// </summary>
public sealed record BookEditMessage(
    string Kind,
    string Program,
    int? Done = null,
    int? Total = null,
    bool? Cancelled = null,
    IReadOnlyList<BookEditRowDto>? Rows = null,
    string? Reason = null);

/// <summary>
/// <c>kind</c>: done | failed | cancelled — how the LLM settings test send ended (Angular ticket 21),
/// sent to the connection that started it. The tokens themselves travel on <c>stream:llm</c>.
/// </summary>
public sealed record LlmTestMessage(string Kind, int ConfigId, string? Reason = null);

/// <summary>
/// <see cref="ProposedEdit"/> on the wire; enums as member names. Shared with the REST reads of a
/// proposal run (<c>BookEditEndpoints</c>) on purpose: a row a client got pushed and a row it read
/// back after missing the push have to be the same shape, or a review screen would need two.
/// </summary>
public sealed record BookEditRowDto(
    string Kind, Guid Id, string DisplayPath, string OldValue, string? NewValue, string Status, string? FailureReason)
{
    public static BookEditRowDto From(ProposedEdit p) =>
        new(p.Kind.ToString(), p.Id, p.DisplayPath, p.OldValue, p.NewValue, p.Status.ToString(), p.FailureReason);
}

/// <summary>Everything a project group member needs on join: current revision plus full status maps.</summary>
public sealed record ProjectSnapshot(
    string Folder,
    long Revision,
    IReadOnlyDictionary<string, NodeStatusSummary?> Nodes,
    int FolderAudioRemaining,
    IReadOnlyDictionary<string, ParagraphStatusEntry?> Paragraphs,
    IReadOnlyDictionary<string, ItemStatusEntry?> Items);

/// <summary>Answer to <c>GetSnapshot</c>: singleton state plus one <see cref="ProjectSnapshot"/> per joined project.</summary>
public sealed record LiveSnapshot(
    QueueMessage Queue,
    AssemblyState Assembly,
    VoiceBatchState VoiceBatch,
    IReadOnlyDictionary<string, string> Watchdog,
    IReadOnlyDictionary<string, string> ServiceStatus,
    ThroughputSnapshot Throughput,
    IReadOnlyDictionary<string, ProjectSnapshot> Projects);

File: src/Read2Me.Web/src/app/live/live-messages.ts
/**
 * Wire shapes of `/hubs/live` (spec D5 "Live hub contract"), mirrored by hand from
 * `Read2Me.App/Live/LiveMessages.cs` + `LiveMessageMapper.cs` + `ILiveClient.cs`. JSON conventions
 * (LiveServiceCollectionExtensions.cs): camelCase properties, enums as member names, nulls omitted
 * — so a nullable C# field is `?: T | null` here. `live-messages.spec.ts` reads the C# sources and
 * fails when a family name or a `kind` string drifts.
 */

import { Guid, NodeStatusSummaryDto, QueueItemStatus } from '@app/api';

// ---- families (ILiveClient.cs [HubMethodName]) --------------------------------------------------

export const LIVE_FAMILIES = [
  'queue',
  'nodeStatus',
  'itemStatus',
  'receipt',
  'assembly',
  'voiceBatch',
  'watchdog',
  'llm',
  'audioGen',
  'throughput',
  'settingsChanged',
] as const;

export type LiveFamily = (typeof LIVE_FAMILIES)[number];

/** Client → server hub methods (LiveHub.cs). */
export const LIVE_HUB_METHODS = {
  joinProject: 'JoinProject',
  leaveProject: 'LeaveProject',
  joinStream: 'JoinStream',
  leaveStream: 'LeaveStream',
  getSnapshot: 'GetSnapshot',
} as const;

export type StreamKind = 'llm' | 'audio';

// ---- kind discriminators (LiveMessageMapper.cs) -------------------------------------------------

export const LIVE_KINDS = {
  assembly: ['phaseStarted', 'progress', 'completed', 'failed', 'cancelled'],
  voiceBatch: ['started', 'progress', 'voiceUpdated', 'completed', 'cancelled'],
  watchdog: ['recoveryStarted', 'containerRestarted', 'serviceHealthy', 'serviceDown'],
  llm: [
    'runStarted',
    'runEnded',
    'requestStarted',
    'delta',
    'streamCompleted',
    'streamFailed',
    'streamAborted',
    'escalationStarted',
  ],
  audioGen: [
    'itemStarted',
    'audioGenerated',
    'normalized',
    'postProcessed',
    'transcribed',
    'verified',
    'failed',
  ],
} as const;

export type AssemblyKind = (typeof LIVE_KINDS.assembly)[number];
export type VoiceBatchKind = (typeof LIVE_KINDS.voiceBatch)[number];
export type WatchdogKind = (typeof LIVE_KINDS.watchdog)[number];
export type LlmKind = (typeof LIVE_KINDS.llm)[number];
export type AudioGenKind = (typeof LIVE_KINDS.audioGen)[number];

// ---- queue (group global, debounced) ------------------------------------------------------------

/** `CharacterQueueService.QueueSnapshot`. */
export interface AttributionQueueState {
  queuedCount: number;
  processingCount: number;
  averageSecondsPerParagraph: number;
  estimatedSecondsRemaining: number;
  completedCount: number;
  currentItemElapsedSeconds: number;
  isBusy: boolean;
}

/** `AudioQueueService.AudioQueueSnapshot`. */
export interface AudioQueueState {
  queuedCount: number;
  processingCount: number;
  averageSecondsPerItem: number;
  estimatedSecondsRemaining: number;
  completedCount: number;
  currentItemElapsedSeconds: number;
}

/** The latched "escalating N items → config (step S)" banner; absent when nothing is escalating. */
export interface EscalationState {
  step: number;
  configName?: string | null;
  itemCount: number;
}

export interface QueueMessage {
  attribution: AttributionQueueState;
  audio: AudioQueueState;
  escalation?: EscalationState | null;
}

// ---- nodeStatus / itemStatus (group project:{folder}, debounced deltas) -------------------------

/** `NodeStatusService.NodeStatusSummary`: the same record `GET /api/projects/{folder}/status` returns. */
export type NodeStatusSummary = NodeStatusSummaryDto;

/** A `null` entry means the node no longer rolls up. */
export interface NodeStatusMessage {
  folder: string;
  nodes: Record<string, NodeStatusSummary | null>;
  folderAudioRemaining: number;
}

export type OutcomeKindName = 'Failed' | 'Unfinished';

export interface OutcomeEntry {
  kind: OutcomeKindName;
  reason?: string | null;
}

export interface ParagraphStatusEntry {
  status?: QueueItemStatus | null;
  outcome?: OutcomeEntry | null;
}

export type AudioReviewState = 'NeedsReview' | 'Dismissed';

/** `AudioReviewService.AudioReviewInfo`. */
export interface AudioReviewInfo {
  state: AudioReviewState;
  normalizeOk: boolean;
  normalizeReason?: string | null;
  verifyOk: boolean;
  wer?: number | null;
  verifyReason?: string | null;
  transcript?: string | null;
  originalTextSnapshot?: string | null;
}

export interface ItemStatusEntry {
  status?: QueueItemStatus | null;
  outcome?: OutcomeEntry | null;
  audioVersion?: number | null;
  review?: AudioReviewInfo | null;
}

/** A `null` entry means the queue forgot the paragraph/item. */
export interface ItemStatusMessage {
  folder: string;
  paragraphs: Record<string, ParagraphStatusEntry | null>;
  items: Record<string, ItemStatusEntry | null>;
}

// ---- receipt (group project:{folder}) -----------------------------------------------------------

export type BookMutationScope = 'Exact' | 'WholeProject';

/** `BookFacets` flag names; the wire value is the comma-joined `[Flags]` string (`"Structure, Audio"`). */
export const BOOK_FACETS = [
  'Structure',
  'ItemText',
  'Attribution',
  'Audio',
  'Reviews',
  'Characters',
  'Narrator',
  'Voices',
  'VoiceRules',
  'ProjectPolicy',
  'NodeTitle',
] as const;

export type BookFacet = (typeof BOOK_FACETS)[number];

/** `"None"`, `"All"`, one facet name, or a comma-separated combination. */
export type BookFacetsValue = string;

export function hasFacet(facets: BookFacetsValue, facet: BookFacet): boolean {
  if (facets === 'All') return true;
  return facets
    .split(',')
    .map((f) => f.trim())
    .includes(facet);
}

export interface BookStructuralRelation {
  kind: 'Split' | 'Merge';
  sourceId: Guid;
  resultId: Guid;
}

export interface BookMutationEffects {
  scope: BookMutationScope;
  facets: BookFacetsValue;
  createdId?: Guid | null;
  nodeIds: Guid[];
  paragraphIds: Guid[];
  paragraphItemIds: Guid[];
  structural: BookStructuralRelation[];
  changedNothing: boolean;
}

export interface ReceiptMessage {
  folder: string;
  /** Verbatim mutation class name, e.g. `CreateCharacterMutation`. */
  mutationName: string;
  mutationId: Guid;
  revision: number;
  effects: BookMutationEffects;
  originId: Guid;
}

/** {@link ReceiptMessage} with the client's own-write flag (from `X-Origin-Id`). */
export interface Receipt extends ReceiptMessage {
  isOwn: boolean;
}

// ---- assembly (group global) --------------------------------------------------------------------

export interface AssemblyMessage {
  kind: AssemblyKind;
  /** `AssemblyPhase` member name; on `phaseStarted`. */
  phase?: string | null;
  /** 0..1; on `progress`. */
  fraction?: number | null;
  /** On `failed`. */
  reason?: string | null;
}

export interface AssemblyState {
  isRunning: boolean;
  currentPhase?: string | null;
  encodePercent: number;
  lastError?: string | null;
  audioRemainingCount: number;
}

// ---- voiceBatch (group global) ------------------------------------------------------------------

export interface VoiceBatchMessage {
  kind: VoiceBatchKind;
  operation?: string | null;
  processed?: number | null;
  total?: number | null;
  failed?: number | null;
  currentVoiceName?: string | null;
  characterId?: Guid | null;
  voiceId?: Guid | null;
  designPrompt?: string | null;
  audioFileName?: string | null;
  transcript?: string | null;
}

export interface VoiceBatchState {
  isRunning: boolean;
  processed: number;
  total: number;
  failed: number;
  currentVoiceName?: string | null;
  currentOperation?: string | null;
  lastError?: string | null;
}

// ---- watchdog (group global) --------------------------------------------------------------------

export interface WatchdogMessage {
  kind: WatchdogKind;
  service: string;
  reason?: string | null;
}

/** Last known watchdog kind per service name. */
export type WatchdogState = Record<string, WatchdogKind | string>;

// ---- llm (group stream:llm) ---------------------------------------------------------------------

export interface LlmMessage {
  kind: LlmKind;
  /** `delta`: a 100 ms batch of thinking text. */
  thinking?: string | null;
  /** `delta`: a 100 ms batch of content text. */
  content?: string | null;
  paragraphPreview?: string | null;
  prompt?: string | null;
  configId?: number | null;
  configName?: string | null;
  tokensIn?: number | null;
  tokensOut?: number | null;
  generationMs?: number | null;
  tokensPerSecond?: number | null;
  reason?: string | null;
  step?: number | null;
  itemCount?: number | null;
}

// ---- audioGen (group stream:audio) --------------------------------------------------------------

export interface AudioGenMessage {
  kind: AudioGenKind;
  id: Guid;
  attempt: number;
  character?: string | null;
  text?: string | null;
  ok?: boolean | null;
  reason?: string | null;
  stepId?: string | null;
  applied?: boolean | null;
  transcript?: string | null;
  wer?: number | null;
  rescued?: boolean | null;
}

// ---- throughput (group global, 1 s while a run is active) ---------------------------------------

export interface ConfigThroughput {
  configId: number;
  configName: string;
  requests: number;
  tokensOut?: number | null;
  generationMs?: number | null;
  tokensPerSecond?: number | null;
}

/** `ThroughputAggregator.ThroughputSnapshot`. */
export interface ThroughputSnapshot {
  hasRun: boolean;
  isRunActive: boolean;
  runThroughput?: number | null;
  perConfig: ConfigThroughput[];
  generationRate?: number | null;
  /** 20 × 500 ms buckets, oldest first; `null` = nothing measured in that bucket. */
  generationRateHistory: (number | null)[];
}

// ---- settingsChanged (group global) -------------------------------------------------------------

export interface SettingsChangedMessage {
  area: string;
}

// ---- snapshots ----------------------------------------------------------------------------------

/** Answer to `JoinProject` and one entry of {@link LiveSnapshot.projects}. */
export interface ProjectSnapshot {
  folder: string;
  revision: number;
  nodes: Record<string, NodeStatusSummary | null>;
  folderAudioRemaining: number;
  paragraphs: Record<string, ParagraphStatusEntry | null>;
  items: Record<string, ItemStatusEntry | null>;
}

/** Answer to `GetSnapshot`. */
export interface LiveSnapshot {
  queue: QueueMessage;
  assembly: AssemblyState;
  voiceBatch: VoiceBatchState;
  watchdog: WatchdogState;
  throughput: ThroughputSnapshot;
  projects: Record<string, ProjectSnapshot>;
}

// ---- family → payload ---------------------------------------------------------------------------

export interface LiveMessageMap {
  queue: QueueMessage;
  nodeStatus: NodeStatusMessage;
  itemStatus: ItemStatusMessage;
  receipt: ReceiptMessage;
  assembly: AssemblyMessage;
  voiceBatch: VoiceBatchMessage;
  watchdog: WatchdogMessage;
  llm: LlmMessage;
  audioGen: AudioGenMessage;
  throughput: ThroughputSnapshot;
  settingsChanged: SettingsChangedMessage;
}
